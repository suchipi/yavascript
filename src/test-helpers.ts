import { spawn, Options as SpawnOptions } from "first-base";
import child_process from "child_process";
import path from "path";
import print from "@suchipi/print";
import * as inspectOptions from "./inspect-options";

export const TMP = child_process
  .execSync("realpath /tmp", { encoding: "utf-8" })
  .trim();

export const rootDir = (...parts: Array<string>) =>
  path.resolve(__dirname, "..", ...parts);

function getBinaryPath(platform: string, arch: string) {
  if (platform === "win32") {
    return path.resolve(__dirname, "..", "bin", "windows", "yavascript.exe");
  } else if (platform === "darwin") {
    if (arch.startsWith("arm")) {
      return path.resolve(__dirname, "..", "bin", "darwin-arm", "yavascript");
    } else {
      return path.resolve(__dirname, "..", "bin", "darwin", "yavascript");
    }
  } else if (platform === "linux") {
    return path.resolve(__dirname, "..", "bin", "linux", "yavascript");
  } else {
    throw new Error("Unsupported platform: " + platform);
  }
}

var binaryPath = getBinaryPath(
  process.platform,
  process.env.YS_TESTS_ARCH || process.arch
);

export { binaryPath, getBinaryPath };

export type EvaluateResult = {
  stdout: string;
  stderr: string;
  code: number | null;
  error: boolean;
};

export async function evaluate(
  code: string,
  options: SpawnOptions = {}
): Promise<EvaluateResult> {
  const runContext = spawn(binaryPath, ["-e", code], options);
  await runContext.completion;
  return runContext.result;
}

export function inspect(value: any): string {
  // options to inspect here match what is given to console.log
  return print(value, inspectOptions.forPrint);
}

export function cleanOutput(input: string): string {
  return input
    .replace(/yavascript:\d+/g, "yavascript")
    .replace(/\/bin\/[^/]+\/yavascript/g, "/bin/.../yavascript")
    .replace(new RegExp(rootDir(), "g"), "<rootDir>");
}

export function cleanResult(input: EvaluateResult): EvaluateResult {
  return {
    ...input,
    stdout: cleanOutput(input.stdout),
    stderr: cleanOutput(input.stderr),
  };
}
