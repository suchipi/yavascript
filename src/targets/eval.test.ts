import { spawn } from "first-base";
import { binaryPath, inspect } from "../test-helpers";

describe("eval target", () => {
  ["-e", "--eval"].forEach((flag) => {
    describe(`via ${flag}`, () => {
      it("evaluates the specified code", async () => {
        const run = spawn(binaryPath, [flag, `console.log("hi")`]);
        await run.completion;
        expect(run.result).toMatchSnapshot();
      });

      it("prints the result when it's not undefined", async () => {
        const run = spawn(binaryPath, [flag, `24`]);
        await run.completion;
        expect(run.result).toMatchSnapshot();
      });

      it("doesn't print anything when the code results in undefined", async () => {
        const run = spawn(binaryPath, [flag, `void 0`]);
        await run.completion;
        expect(run.result).toMatchSnapshot();
      });

      it("prints non-empty completion value", async () => {
        const run = spawn(binaryPath, [flag, `"hello"; var a = 5;`]);
        await run.completion;
        expect(run.result).toMatchSnapshot();
      });

      it("can evaluate coffeescript", async () => {
        const run = spawn(binaryPath, [
          flag,
          `console.log "hi"`,
          "--lang",
          "coffeescript",
        ]);
        await run.completion;
        expect(run.result).toMatchSnapshot();
      });

      it("can evaluate jsx", async () => {
        const run = spawn(binaryPath, [
          flag,
          `<div><a key="hi" href="#" /><></></div>`,
          "--lang",
          "jsx",
        ]);
        await run.completion;
        expect(run.result).toMatchSnapshot();
      });

      ["ts", "typescript"].forEach((lang) => {
        it(`can evaluate typescript (via --lang ${lang})`, async () => {
          const run = spawn(binaryPath, [
            flag,
            `(function something<T>(blah: number): T { return 5 as any; }).name`,
            "--lang",
            lang,
          ]);
          await run.completion;
          expect(run.result).toMatchSnapshot();
        });
      });

      it("can evaluate tsx", async () => {
        const run = spawn(binaryPath, [
          flag,
          `(function something(blah: number) { return <div><a key="hi" href="#" /><></></div> })()`,
          "--lang",
          "tsx",
        ]);

        await run.completion;
        expect(run.result).toMatchSnapshot();
      });
    });
  });
});
