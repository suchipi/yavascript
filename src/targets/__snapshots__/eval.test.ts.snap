// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eval target via --eval can evaluate coffeescript 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hi
",
}
`;

exports[`eval target via --eval can evaluate jsx 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: \\"div\\"
  props: {
    children: [
      {
        $$typeof: Symbol(JSX.Element)
        type: \\"a\\"
        props: {
          key: \\"hi\\"
          href: \\"#\\"
        }
        key: \\"hi\\"
      }
      {
        $$typeof: Symbol(JSX.Element)
        type: Symbol(JSX.Fragment)
        props: null
        key: null
      }
    ]
  }
  key: null
}
",
}
`;

exports[`eval target via --eval can evaluate tsx 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: \\"div\\"
  props: {
    children: [
      {
        $$typeof: Symbol(JSX.Element)
        type: \\"a\\"
        props: {
          key: \\"hi\\"
          href: \\"#\\"
        }
        key: \\"hi\\"
      }
      {
        $$typeof: Symbol(JSX.Element)
        type: Symbol(JSX.Fragment)
        props: null
        key: null
      }
    ]
  }
  key: null
}
",
}
`;

exports[`eval target via --eval can evaluate typescript (via --lang ts) 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "something
",
}
`;

exports[`eval target via --eval can evaluate typescript (via --lang typescript) 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "something
",
}
`;

exports[`eval target via --eval doesn't print anything when the code results in undefined 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`eval target via --eval evaluates the specified code 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hi
",
}
`;

exports[`eval target via --eval prints non-empty completion value 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hello
",
}
`;

exports[`eval target via --eval prints the result when it's not undefined 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "24
",
}
`;

exports[`eval target via -e can evaluate coffeescript 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hi
",
}
`;

exports[`eval target via -e can evaluate jsx 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: \\"div\\"
  props: {
    children: [
      {
        $$typeof: Symbol(JSX.Element)
        type: \\"a\\"
        props: {
          key: \\"hi\\"
          href: \\"#\\"
        }
        key: \\"hi\\"
      }
      {
        $$typeof: Symbol(JSX.Element)
        type: Symbol(JSX.Fragment)
        props: null
        key: null
      }
    ]
  }
  key: null
}
",
}
`;

exports[`eval target via -e can evaluate tsx 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: \\"div\\"
  props: {
    children: [
      {
        $$typeof: Symbol(JSX.Element)
        type: \\"a\\"
        props: {
          key: \\"hi\\"
          href: \\"#\\"
        }
        key: \\"hi\\"
      }
      {
        $$typeof: Symbol(JSX.Element)
        type: Symbol(JSX.Fragment)
        props: null
        key: null
      }
    ]
  }
  key: null
}
",
}
`;

exports[`eval target via -e can evaluate typescript (via --lang ts) 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "something
",
}
`;

exports[`eval target via -e can evaluate typescript (via --lang typescript) 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "something
",
}
`;

exports[`eval target via -e doesn't print anything when the code results in undefined 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`eval target via -e evaluates the specified code 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hi
",
}
`;

exports[`eval target via -e prints non-empty completion value 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hello
",
}
`;

exports[`eval target via -e prints the result when it's not undefined 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "24
",
}
`;
